services/src/pro/contexts/mod.rs

```diff
@@ -137,6 +183,30 @@ where
         op
     }
 
+    async fn create_raster_query_processor(
+        &self,
+        operator: Box<dyn InitializedRasterOperator>,
+    ) -> geoengine_operators::util::Result<TypedRasterQueryProcessor> {
+        check_quota(&*self.user_db, &self.session).await?;
+        operator.query_processor()
+    }
+
+    async fn create_vector_query_processor(
+        &self,
+        operator: Box<dyn InitializedVectorOperator>,
+    ) -> geoengine_operators::util::Result<TypedVectorQueryProcessor> {
+        check_quota(&*self.user_db, &self.session).await?;
+        operator.query_processor()
+    }
+
+    async fn create_plot_query_processor(
+        &self,
+        operator: Box<dyn InitializedPlotOperator>,
+    ) -> geoengine_operators::util::Result<TypedPlotQueryProcessor> {
+        check_quota(&*self.user_db, &self.session).await?;
+        operator.query_processor()
+    }
```

Comment:
Is this the right place?
Shouldn't it be checked when we call `query`?

Reply:
During initialization of the query processor we already do some work, like computing ResultDescriptors, initializing models and idk what. So why would we want to do that work and only abort later when finally query is called?

Reply:
Wouldn't you want to run a query as long as you have resources and not to be prevented to run the first query although you would have had resources at least for that?

Reply:
that's how it currently works: If you have at least 1 credit, you can create the query processor and query it as often as you want.

If we want to check every query-call, we have to do things differently. Like introducing a wrapper operator that checks this in the query-method before delegating it to the real workflow.

Reply:
ok so I built it into the query-method now. We ultimately need it there, because otherwise one could create a processor with 1 credit and use it forever because the quota is not checked again.

So right now, every processor in the chain will check the quota. We discussed caching it instead. The cache would be in the query context. Right now, this is not possible because the query context can be re-used for subsequent queries.

I would propose to make the QueryContext non-reusable. So it should be consumed by the query-method. Then we can cache the quota there, and, we can also have something like a unique query id stored in the context that we can use for stuff like logging and quota tracking.

In order for the QueryContext to be consumed, we have to change the parameter in the query method from a fat pointer to a box though. And: for operators with more than one child, we would need to clone the QueryContext, in oder to put it into each child's query method.

Reply:
I guess it is already designed this way, so we could do that.

Reply:
the problem here is, if the QueryContext is cloneable, there is no benefit in it being consumed by the query method. You can just clone it before you put it in. So we would somehow have to invalidate it after being put into a query-method, but put it in again, for all the child processors.

I guess then the best idea would be to avoid wrapping every single processor with the quota check, but only wrap the top-level one. But how can the execution context distinguish the top-level operator from the other ones? Maybe we have to do the top-level wrapping in a different place.

One idea would be to do it in the `Workflow`. Right now, we access the top-level operator by `workflow.operator` directly. We could make `Workflow` into a trait with a method `operator()` which, in free just outputs the operator but in pro it wraps it with a quota checker.